### PR TITLE
Databricks supports horovod.spark.estimator

### PR DESCRIPTION
**Pull request by liangz1**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

* Add `DBFSLocalStore` that allows horovod.spark Estimator to use Databricks Filesystem local files API to persist intermediate data and training artifacts.
* Update the ways to pin GPU: when GPU-aware scheduling is available, pin GPU to the assigned GPU index from spark; otherwise, pin GPU to `hvd.local_rank()`.

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

File: horovod/spark/keras/remote.py [hunk 1]
```diff
@@ -301,15 +305,25 @@ def fn(hvd, tf, keras):
         for gpu in gpus:
             tf.config.experimental.set_memory_growth(gpu, True)
         if gpus:
-            tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], 'GPU')
+            if os.environ.get('CUDA_VISIBLE_DEVICES') is None:
```
Comment (liangz1) [inline horovod/spark/keras/remote.py:308]: In the test(non-databricks environment), assert this branch is met.

Comment (WeichenXu123) [reply horovod/spark/keras/remote.py:308]: What's the value of CUDA_VISIBLE_DEVICES set in non-databricks environment?

**Comment by WeichenXu123** [comment]
Two issues: @mengxr 

* When driver fetch result model from rank 0 process, current impl is rank 0 serialize model to bytes array and sent back, but now the model already saved into DBFS, should we optimize it by "rank 0 process sent model saved dbfs path to driver, and driver load model from the dbfs path ?"

* Should we implement the DBFSLocalStore by:
  saving model to local file system temporary and then copy local file to dbfs by `sync_fn` interface ? This is the way HDFSStore used and it can support H5 model format.

**Comment by liangz1** [comment]
1. For the first issue:
Current impl: 
Pros:
(1) Keep the return value type of the train function the same as LocalStore and HDFSStore. The code change is minimal.
Cons:
(1) Not taking the advantage of DBFS FUSE feature. But this is as good as the existing solution.

Return DBFS Path:
Pros:
(1) Replace "load & serialize model on worker + pickle to the driver" with "pickle path to driver + load model on the driver". Save the pickling cost.
Cons:
(1) Need to add if-else branches to handle different cases. More code change. 
(2) I'm not sure whether there is a consistency issue: After a file is written on worker, whether it is guaranteed to be available immediately on driver.

2. The second issue is discussed in the design doc "design decision 1: option 4". In short, the cons are more complex code change and unnecessary copy overhead.

File: horovod/spark/torch/remote.py [hunk 2]
```diff
@@ -119,7 +121,14 @@ def train(serialized_model, optimizer_cls, model_opt_state_serialized,
...
         if cuda_available:
             # Horovod: pin GPU to local rank.
-            torch.cuda.set_device(hvd.local_rank())
+            if os.environ.get('CUDA_VISIBLE_DEVICES') is None:
+                assert not is_dbfs
+                torch.cuda.set_device(hvd.local_rank())
+            else:
+                # Databricks pyspark sets CUDA_VISIBLE_DEVICES for GPU scheduling.
+                # Pin the only visible GPU allocated to this task.
+                assert is_dbfs
```
Comment (liangz1) [inline horovod/spark/torch/remote.py:130]: This will fail in CI. We should use is_dbfs to determine whether to use pin to GPU[0]

File: test/test_spark.py [hunk 3]
```diff
@@ -1669,3 +1669,46 @@ def test_to_list(self):
...
+        dbfs_ckpt_path = dbfs_store.prefix_path + "/" + dbfs_ckpt_name
+        model.save(dbfs_ckpt_path)
+        serialized_model_dbfs = dbfs_store.read_serialized_keras_model(dbfs_ckpt_path)
+        reconstructed_model_dbfs = deserialize_keras_model(serialized_model_dbfs)
+        assert reconstructed_model_dbfs.get_config() == model.get_config()
+        assert reconstructed_model_dbfs.get_weights() == model.get_weights()
+
+        # test local_store.read_serialized_keras_model
+        local_ckpt_path = local_store.prefix_path + "/" + local_store.get_checkpoint_filename()
+        model.save(local_ckpt_path)
```
Comment (liangz1) [inline test/test_spark.py:1707]: For TF 1.15.0, SavedModel format is not selected by adding the ".tf" suffix. https://github.com/tensorflow/docs/blob/920057b528427ac647b8dcb93d898fce76507d3d/site/en/r1/guide/saved_model.md#build-and-load-a-savedmodel The syntax for saving in SavedModel is quite different from the current one.
Manual test on databricks with tf downgraded to 1.15 also fails with the h5py error.

File: horovod/spark/torch/remote.py [hunk 4]
```diff
@@ -119,7 +121,14 @@ def train(serialized_model, optimizer_cls, model_opt_state_serialized,
...
+                torch.cuda.set_device(0)
```
Comment (mengxr) [inline horovod/spark/torch/remote.py:129]: We should do this:

~~~
if has GPU-aware scheduling:
  pin to the assigned GPU index no matter whether CUDA_VISIBLE_DEVICES is defined or not
else:
  pin to local rank (this might have issues)
~~~

File: test/test_spark.py (hunk 3 above)
Comment (liangz1) [reply test/test_spark.py:1707]: https://dogfood.staging.cloud.databricks.com/#notebook/29305617/command/29305633

File: horovod/spark/common/store.py [hunk 5]
```diff
@@ -164,6 +165,15 @@ def read(self, path):
         with self.get_filesystem().open(self.get_localized_path(path), 'rb') as f:
             return f.read()
 
+    def read_serialized_keras_model(self, ckpt_path, model=None):
+        """
+        This function will be overridden by DBFSLocalStore.
```
Comment (mengxr) [inline horovod/spark/common/store.py:170]: * First sentence should describe what this method does.
* Describe the params and the return value.

File: horovod/spark/common/store.py [hunk 6]
```diff
@@ -164,6 +165,15 @@ def read(self, path):
...
+        """
+        from horovod.runner.common.util import codec
+
+        model_bytes = self.read(ckpt_path)
+        return codec.dumps_base64(model_bytes)
```
Comment (mengxr) [inline horovod/spark/common/store.py:175]: beyond this PR: base64 is not very efficient in encoding bytes. Why not transferring the model bytes directly? cc: @tgaddair

File: horovod/spark/common/store.py [hunk 7]
```diff
@@ -431,3 +441,40 @@ def _check_url(self, url, prefix, path):
     @classmethod
     def filesystem_prefix(cls):
         return cls.FS_PREFIX
+
+
+class DBFSLocalStore(LocalStore):
+    """Uses DBFS as a store of intermediate data and training artifacts.
```
Comment (mengxr) [inline horovod/spark/common/store.py:447]: Databricks File System (DBFS) local file APIs

File: horovod/spark/common/store.py [hunk 8]
```diff
@@ -431,3 +441,40 @@ def _check_url(self, url, prefix, path):
...
+
+    Initialized from a `prefix_path` starts with `/dbfs/...`, see
+    https://docs.databricks.com/data/databricks-file-system.html#local-file-apis.
+
+    Note that DBFSLocalStore cannot be created via Store.create("/dbfs/..."), which will create a
```
Comment (mengxr) [inline horovod/spark/common/store.py:452]: Maybe a follow-up: If user calls `Store.create("dbfs:/")`, shall we convert it into DBFSLocalStore?

File: horovod/spark/common/store.py [hunk 9]
```diff
@@ -431,3 +441,40 @@ def _check_url(self, url, prefix, path):
...
+        Returns serialized keras model. On Databricks, only TFKeras is supported, not BareKeras.
+        The parameter `model` is for providing the model structure when the checkpoint file only
+        contains model weights.
+        """
+        import tensorflow
+        from tensorflow import keras
+        from horovod.spark.keras.util import TFKerasUtil
+
+        if LooseVersion(tensorflow.__version__) < LooseVersion("2.0.0"):
+            model.load_weights(ckpt_path)
```
Comment (mengxr) [inline horovod/spark/common/store.py:477]: When `model=None`? If we never expect it to happen, we shouldn't leave the default to None.

File: horovod/spark/common/util.py [hunk 10]
```diff
@@ -695,3 +695,13 @@ def to_list(var, length):
                              "label_cols")
 
     return var
+
+
+def _get_allocated_gpu(hvd):
```
Comment (mengxr) [inline horovod/spark/common/util.py:700]:
~~~python
_get_assigned_gpu_or_default(default):
~~~

File: horovod/spark/keras/remote.py [hunk 11]
```diff
@@ -154,7 +157,11 @@ def train(serialized_model, train_rows, val_rows, avg_row_size):
                 if _checkpoint_callback:
                     _checkpoint_callback.filepath = ckpt_file
                 else:
-                    _checkpoint_callback = k.callbacks.ModelCheckpoint(ckpt_file)
+                    if is_dbfs and LooseVersion(tf.__version__) < LooseVersion("2.0.0"):
+                        _checkpoint_callback = k.callbacks.ModelCheckpoint(ckpt_file,
```
Comment (mengxr) [inline horovod/spark/keras/remote.py:161]: would be nice to have an inline comment to explain why (DBFS local APIs does not support random write).

**Review by mengxr — CHANGES_REQUESTED** [review]

File: horovod/spark/common/store.py (hunk 8 above)
Comment (liangz1) [reply horovod/spark/common/store.py:452]: Added to backlog.

File: test/test_spark.py [hunk 12]
```diff
@@ -1669,3 +1669,55 @@ def test_to_list(self):
...
+        with get_dbfs_output_dir() as run_output_dir:
+            dbfs_ckpt_path = run_output_dir + "/" + local_store.get_checkpoint_filename()
+            if LooseVersion(tensorflow.__version__) < LooseVersion("2.0.0"):
+                model.save_weights(dbfs_ckpt_path)
+            else:
+                model.save(dbfs_ckpt_path)
+            serialized_model_dbfs = dbfs_store.read_serialized_keras_model(dbfs_ckpt_path, model)
+            reconstructed_model_dbfs = deserialize_keras_model(serialized_model_dbfs)
+            if LooseVersion(tensorflow.__version__) >= LooseVersion("2.3.0"):
+                assert reconstructed_model_dbfs.get_config() == model.get_config()
```
Comment (liangz1) [inline test/test_spark.py:1710]: In TF 2.2.0, pytest fails with
~~~
assertion error: {'build_input_shape': TensorShape([None, 32])} != {'build_input_shape': TensorShape([None, 32])}
~~~
In TF 2.1.0, 2.0.0, it has
~~~
{..., 'batch_input_shape': [None, 32]} != {..., 'batch_input_shape': (None, 32)}
~~~
but pytest did not fail with assertion error. 

Thus I disabled the model equality check for versions < 2.3.0.

File: docs/spark.rst [hunk 13]
```diff
@@ -290,5 +290,29 @@ Environment knobs
 -----------------
 * ``HOROVOD_SPARK_START_TIMEOUT`` - sets the default timeout for Spark tasks to spawn, register, and start running the code.  If executors for Spark tasks are scheduled on-demand and can take a long time to start, it may be useful to increase this timeout on a system level.
 
+Horovod in Spark on Databricks
```
Comment (mengxr) [inline docs/spark.rst:293]: Shall we just call it "Horovod on Databricks"?

File: docs/spark.rst [hunk 14]
```diff
@@ -290,5 +290,29 @@ Environment knobs
...
+------------------------------
+Horovod in Spark in supported on
+`Databricks Runtime 7.4 ML <https://docs.databricks.com/release-notes/runtime/releases.html>`__
```
Comment (mengxr) [inline docs/spark.rst:296]: Do not mention a future Runtime version here. Instead, mention the min Horovod version to have DBFSLocalStore or just don't mention it at all.

File: docs/spark.rst [hunk 15]
```diff
@@ -290,5 +290,29 @@ Environment knobs
...
+and above. To run Horovod in Spark on Databricks, use `DBFSLocalStore` as the store object:
+
+.. code-block:: python
+
+    store = DBFSLocalStore(prefix_path='/dbfs/...')
+
+The `DBFSLocalStore` uses `Databricks File System (DBFS) local file APIs <https://docs.databricks.com/data/databricks-file-system.html#local-file-apis>`__
```
Comment (mengxr) [inline docs/spark.rst:303]: Could you add two links one for AWS and one for Azure?

File: docs/spark.rst [hunk 16]
```diff
@@ -290,5 +290,29 @@ Environment knobs
...
+as a store of intermediate data and training artifacts.
+
+Databricks preconfigures GPU-aware scheduling for you on GPU clusters. See the
```
Comment (mengxr) [inline docs/spark.rst:306]: I don't think Databricks setup is any different from a Spark cluster with GPU scheduling enabled. We can just say Databricks pre-configures GPU-aware scheduling on Databricks Runtime 7.0 ML GPU and above, and link to GPU scheduling instructions.

File: horovod/spark/common/util.py [hunk 17]
```diff
@@ -695,3 +695,13 @@ def to_list(var, length):
                              "label_cols")
 
     return var
+
+
+def _get_assigned_gpu_or_default(default):
+    try:
+        # if GPU-aware scheduling is available, pin to the assigned GPU index
+        from horovod.spark.task import get_available_devices
+        return int(get_available_devices()[0])
```
Comment (mengxr) [inline horovod/spark/common/util.py:704]: What exception do you expect this call to raise? If it returns an empty list and `[0]` throws an index out of bound error, then it is better to do:

~~~python
from horovod.spark.task import get_available_devices
available_devices = get_available_devices()
if available_devices:
  return int(available_devices[0])
else:
  return default
~~~

The difference is that if the cluster has GPU-aware configured but GPU index was done wrongly, e.g., "gpu1" instead of "1". We should throw an exception instead of fallback.

**Review by mengxr — CHANGES_REQUESTED** [review]
Some minor comments.

**Review by mengxr — APPROVED** [review]
LGTM. @tgaddair Could you take a look?

File: horovod/spark/common/store.py (hunk 6 above)
Comment (tgaddair) [reply horovod/spark/common/store.py:179]: The original intent was to be able to store as part of a JSON file that encoded Estimator params, so we would represent the model as a base64 encoded string.  I believe this was changed, however, so may not be needed anymore.

@abditag2 can confirm.

File: horovod/spark/common/store.py [hunk 18]
```diff
@@ -431,3 +445,41 @@ def _check_url(self, url, prefix, path):
...
+        super(DBFSLocalStore, self).__init__(prefix_path, *args, **kwargs)
+
+    def get_checkpoint_filename(self):
+        # Use the default Tensorflow SavedModel format in TF 2.x. In TF 1.x, the SavedModel format
+        # is used by providing `save_weights_only=True` to the ModelCheckpoint() callback.
+        return 'checkpoint.tf'
+
+    def read_serialized_keras_model(self, ckpt_path, model):
+        """
+        Returns serialized keras model. On Databricks, only TFKeras is supported, not BareKeras.
```
Comment (tgaddair) [inline horovod/spark/common/store.py:473]: Note: we will probably be dropping BareKeras support in the next release as well.

File: horovod/spark/common/store.py [hunk 19]
```diff
@@ -431,3 +445,41 @@ def _check_url(self, url, prefix, path):
...
+        The parameter `model` is for providing the model structure when the checkpoint file only
+        contains model weights.
+        """
+        import tensorflow
```
Comment (tgaddair) [inline horovod/spark/common/store.py:477]: Is the reason for having this method unique to this store that it will not work with BareKeras?  If so, we can move this out once we drop BareKeras support.

File: horovod/spark/keras/remote.py [hunk 20]
```diff
@@ -154,7 +157,14 @@ def train(serialized_model, train_rows, val_rows, avg_row_size):
                 if _checkpoint_callback:
                     _checkpoint_callback.filepath = ckpt_file
                 else:
-                    _checkpoint_callback = k.callbacks.ModelCheckpoint(ckpt_file)
+                    if is_dbfs and LooseVersion(tf.__version__) < LooseVersion("2.0.0"):
```
Comment (tgaddair) [inline horovod/spark/keras/remote.py:160]: Similar to previous comment: we should be able to adopt this pattern universally after we drop BareKeras, without needing to special-case DBFS.

**Review by tgaddair — COMMENTED** [review]
Just to check my understanding: all of the special handling of model checkpointing for DBFS is specific to the fact that you are only supporting TensorFlow-native Keras, is that correct?  My assumption here is that we can remove this special-casing once we drop support for standalone Keras, but I want to make sure that's correct.

**Comment by liangz1** [comment]
@tgaddair all of the special handling of model checkpointing for DBFS is to avoid using h5 format, which requires random writes, and DBFS does not support that.

**Comment by tgaddair** [comment]
Ah, gotcha @liangz1.  Is there any reason not to use the SavedModel format over H5?  Are you aware of any performance differences?  I ask, because it would simplify things to converge on a single format.  If we need to support both, than it would be cleaner to have the store specify the model checkpoint format, then have separate handlers for each format (instead of special-casing on the store type).

**Comment by liangz1** [comment]
One issue with SavedModel is we have to use save_weights_only in TF 1.x to enabled it, see [comment above](https://github.com/horovod/horovod/pull/2234#discussion_r483703692). That makes the code looks ugly.

Performance-wise: I searched online and didn't see discussions about performance regressions. I noticed that TF 2.x is default at SavedModel format and refers to h5 as "the old format". I would support unifying to SavedModel format in all cases, which could be done in follow-up PRs.

File: horovod/spark/common/store.py (hunk 6 above)
Comment (abditag2) [reply horovod/spark/common/store.py:179]: Correct. When we serialize the estimator, the model is stored as a b64 encoded string in Json along with the rest of the parameters of the SparkMLEstimator.

Comment (tgaddair) [reply horovod/spark/common/store.py:179]: Thanks, so we may not need it here then since we're reading from disk, but this can be addressed in a follow-up PR.

**Review by tgaddair — APPROVED** [review]
Let's go ahead and merge to unblock you, and we can do some refactoring in a follow-up PR.

**Comment by tgaddair** [comment]
Thanks for the PR, @liangz1!

File: horovod/spark/common/store.py (hunk 6 above)
Comment (abditag2) [reply horovod/spark/common/store.py:179]: Yes.

**Comment by liangz1** [comment]
Thanks for the timely review! 🎉 @tgaddair @abditag2
